src/System.Security.Cryptography.Algorithms/src/Internal/Cryptography/RandomNumberGeneratorImplementation.cs

```diff
@@ -0,0 +1,58 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+namespace System.Security.Cryptography
+{
+    internal sealed partial class RandomNumberGeneratorImplementation : RandomNumberGenerator
+    {
+        public override void GetNonZeroBytes(byte[] data)
+        {
+            if (data == null)
+                throw new ArgumentNullException(nameof(data));
+
+            GetBytes(data);
+
+            int indexOfFirst0Byte = data.Length;
+            for (int i = 0; i < data.Length; i++)
+            {
+                if (data[i] == 0)
+                {
+                    indexOfFirst0Byte = i;
+                    break;
+                }
+            }
+
+            for (int i = indexOfFirst0Byte; i < data.Length; i++)
+            {
+                if (data[i] != 0)
+                {
+                    data[indexOfFirst0Byte++] = data[i];
+                }
+            }
+
+            while (indexOfFirst0Byte < data.Length)
+            {
+                // This should be more than enough to fill the rest in one iteration
+                byte[] tmp = new byte[2 * (data.Length - indexOfFirst0Byte)];
```

Comment:
Rather than allocating more space than we need, how about adding a private GetBytes(byte[], int) overload, and have that be the one we define in each partial, e.g.
```C#
private void GetBytes(byte[] data, int dataLength)
{
    Debug.Assert(data.Length > 0);
    if (!Interop.Crypto.GetRandomBytes(data, dataLength))
    {
        throw Interop.Crypto.CreateOpenSslCryptographicException();	
    }
}
```
Then put the primary overload in the main partial, e.g.
```C#
public sealed override void GetBytes(byte[] data)
{
    ValidateGetBytesArgs(data);
    if (data.Length > 0)
    {
        GetBytes(data, data.Length);
    }
}
```
and then here doing a single allocation outside of the loop, of the right initial length, and refilling it if need be, e.g.
```C#
int bytesRemaining = data.Length - indexOfFirst0Byte;
if (bytesRemaining > 0)
{
    var tmp = new byte[bytesRemaining];
    while (bytesRemaining > 0)
    {
        GetBytes(tmp, bytesRemaining);
        for (int i = 0; i < bytesRemaining; i++)
        {
            if (tmp[i] != 0)
            {
                data[indexOfFirst0Byte++] = tmp[i];
                if (indexOfFirst0Byte >= data.Length) break;
            }
        }
        bytesRemaining = data.Length - indexOfFirst0Byte;
    }
}
```
or something along those lines.

Actually, do we even need to allocate at all?  What if the whole method were instead like:
```C#
private static void GetNonZeroBytes(byte[] data)
{
    ValidateGetBytesArgs(data);
    if (data.Length > 0)
    {
        int offset = 0;
        while (offset < data.Length)
        {
            // Fill the remaining portion of the array with random bytes.
            GetBytes(data, offset, data.Length - offset);

            // Find the first zero in the remaining portion.
            int indexOfFirst0Byte = data.Length;
            for (int i = offset; i < data.Length; i++)
            {
                if (data[i] == 0)
                {
                    indexOfFirst0Byte = i;
                    break;
                }
            }

            // If there were any zeros, shift down all non-zeros,
            // leaving all of the zeros at the end.
            for (int i = indexOfFirst0Byte; i < data.Length; i++)
            {
                if (data[i] != 0)
                {
                    data[indexOfFirst0Byte++] = data[i];
                }
            }

            // Go again on the space filled with zeros, if any.
            offset = indexOfFirst0Byte;
        }
    }
}
```
This would necessitate changing that other overload of GetBytes to be ```GetBytes(byte[] data, int offset, int length)```, and it would necessitate changing the P/Invokes to accept a ```byte*``` rather than a ```byte[]```.